52 implement new energy concept grid strength

Nullformed opened the pull request.

Nullformed committed:
Implemented Grid Strength

Nullformed committed:
UI label text changes

Nullformed committed:
Energy as multiplier

r-zander commented on js/classes.js:
```diff
@@ -8,6 +8,8 @@ class EffectType {
     static Population = new EffectType('x', 'Population');
     static Energy = new EffectType('+', 'Energy');
+    static EnergyFactor = new EffectType('x', 'Energy');
+    static GridStrength = new EffectType('+', 'GridStrength');
```
ist GridStrength nicht ein Task, kein Effect?

r-zander replied:
ah ich sehe es gibt auch nen Task der so heißt. puh. Ich glaube den EffectType braucht's dann einfach nicht

r-zander commented on js/gameConfig.js:
```diff
@@ -24,6 +24,8 @@ const units = {
     storedEnergy: 'Wh'
 };
 
+const gridStrength = new GridStrength({name:'GridStrength', title: 'Grid Strength', maxXp: 100, effects: [{effectType: EffectType.GridStrength, baseValue: 1}]});
```
okay ... yeah. 🤯 GridStrength hat einfach selbst nen Effect der GridStrength angibt. Bin mir nicht sicher ob das genial oder wahnsinnig ist.
Also da ja eigentlich die Ansage war; Level von GridStrength = GridStrength würde ich das wohl auch bei dieser simplen Logik lassen, ohne extra Effect Layer

Nullformed replied:
so funktionieren tasks jetzt halt... ohne effectType kein effect. 💯

r-zander replied:
Ja aber GridStrength ist halt auch kein Effect am Ende des Tages. Das is nen Level von nem Task. Nen Task den man zwar nicht deaktivieren kann, aber halt progressed.
Hast ja sogar die Methode `GridStrength.prototype.getGridStrength` die das Level verwendet.

r-zander commented on js/gameConfig.js:
```diff
@@ -34,20 +36,20 @@ const skillBaseData = {
-    NanoFertilityDrones: new ModuleOperation({title: 'Nano-Fertility Drones', effects: [{effectType: EffectType.Population, baseValue: 5}], maxXp: 400, energyConsumption: 50}),
-    HoloCommunityHub: new ModuleOperation({title: 'Holo-Community Hub', effects: [{effectType: EffectType.Population, baseValue: 5}], maxXp: 400, energyConsumption: 60}),
-    TemporalBreedingPods: new ModuleOperation({title: 'Temporal Breeding Pods', effects: [{effectType: EffectType.Population, baseValue: 5}], maxXp: 400, energyConsumption: 80}),
+    Garbage: new ModuleOperation({title: 'Garbage', effects: [{effectType: EffectType.Population, baseValue: 5}, {effectType: EffectType.Energy, baseValue: 1}], maxXp: 400, energyConsumption: 1}),
+    Diesel: new ModuleOperation({title: 'Diesel', effects: [{effectType: EffectType.Population, baseValue: 5}, {effectType: EffectType.Energy, baseValue: 2}], maxXp: 50, energyConsumption: 1}),
+    Plastics: new ModuleOperation({title: 'Plastics', effects: [{effectType: EffectType.Population, baseValue: 5}, {effectType: EffectType.Energy, baseValue: 1}], maxXp: 100, energyConsumption: 1}),
+    Steel: new ModuleOperation({title: 'Steel', effects: [{effectType: EffectType.Population, baseValue: 5}, {effectType: EffectType.EnergyFactor, baseValue: 2}], maxXp: 200, energyConsumption: 1}),
```
EnergyFactor hier macht nen guten Eindruck 👍 
ich würde allerdings gerne noch `energyConsumption` umbenannt sehen in `gridLoad`

r-zander commented on js/main.js:
```diff
@@ -741,32 +742,47 @@ function doTasks() {
     // Legacy
     doTask(gameData.currentSkill);
     doTask(gameData.currentBattle);
+    doTask(gridStrength);
```
muss das nicht auch iwie in die gameData rein, weils sonst nicht gespeichert wird?

Nullformed replied:
ja, die gameData hat halt auch nur die referenz aufs selbe singleton. daher so schöner?

r-zander commented on js/main.js:
```diff
@@ -786,17 +802,12 @@ function getEnergyUsage() {
-    gameData.storedEnergy += generated;
-    const usage = applySpeed(getEnergyUsage());
-    gameData.storedEnergy -= usage;
-    if (gameData.storedEnergy < 0) {
+    if (getNet() < 0) {
         goBlackout();
     }
 }
```
ich glaub die ganzen Methoden hier rund um Energy kannst du einfach wegschmeißen. es sollte nur 3 globale methoden geben:
- getGridStrength
- getGridLoad
- getRemainingGridStrength

alles andere interessiert nur den GridStrength Task intern

r-zander commented on js/main.js:
```diff
@@ -1271,6 +1282,11 @@ function loadGameData() {
         replaceSaveDict(gameData.itemData, gameDataSave.itemData);
         replaceSaveDict(gameData.battleData, gameDataSave.battleData);
 
+        gameData.gridStrength.level = gameDataSave['gridStrength'].level;
+        gameData.gridStrength.maxLevel = gameDataSave['gridStrength'].maxLevel;
+        gameData.gridStrength.xp = gameDataSave['gridStrength'].xp;
+        gameDataSave['gridStrength'] = gameData.gridStrength;
+
```
neee, leg gridStrength da einfach rein so wie currentJob oder currentSkill da angelegt waren

Nullformed replied:
die dinger wären halt in taskData. die gridStrength nicht. aber ich schau mal...

r-zander commented on js/main.js:
```diff
@@ -1400,6 +1416,7 @@ function initStationName() {
 }
 
 function setDefaultCurrentValues() {
+    gameData.gridStrength = gridStrength;
```
das sind hier halt nicht nur die Daten sondern auch der ganze Task incl level und XP, oder?

Nullformed replied:
naja, die reference auf den Task.

r-zander commented on js/vfx.js:
```diff
@@ -208,7 +208,11 @@ GameEvents.TaskLevelChanged.subscribe(function (taskInfo) {
         taskProgressBar = getBattleElement(taskInfo.name);
-    } else {
+    }
+    else if (taskInfo.type === 'gridstrength') {
```
dieses .toLowerCase alter 🙈 ich muss das mal fixen, das' scheiße

Nullformed committed:
Renamed energyConsumption to gridLoad

Nullformed committed:
Introduced Task assignSaveData method.
Grid Strength moved to taskData.
Introduced function to assign & replace saveDict task object save data.

Nullformed committed:
Replaced grid strength effect with  just 'level' as base: https://github.com/r-zander/progress-station/pull/9/files#r1353277400

r-zander commented on js/classes.js:
```diff
@@ -301,14 +306,18 @@ class ModuleOperation extends Job {
 
     getGridStrength(){
-         return this.getEffect(EffectType.GridStrength) + Math.floor(this.maxLevel / 10);
+         return this.level + Math.floor(this.maxLevel / 10);
```
wie kommst du hier auf `this.maxLevel / 10` ? Das ist doch "nur" für die generated Energy relevant

r-zander commented on js/main.js:
```diff
@@ -599,7 +599,7 @@ function updateItemRows() {
         const row = document.getElementById('row ' + item.name);
         const button = row.getElementsByClassName('button')[0];
-        button.disabled = getNet() < item.getEnergyUsage();
+        button.disabled = getRemainingGridStrength() < item.getEnergyUsage();
```
damit wäre das dann item.getGridLoad()?

Nullformed committed:
Merge branch 'master' into 52-implement-new-energy-concept-grid-strength

# Conflicts:
#	index.html
#	js/classes.js
#	js/gameConfig.js
#	js/main.js

Nullformed committed:
Removed wrong grid strength max level bonus.

Nullformed committed:
Refactored getEnergyUsage to getGridLoad
